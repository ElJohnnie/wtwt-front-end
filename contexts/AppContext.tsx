import React, {
  createContext,
  useState,
  useRef,
  useCallback,
  ReactNode,
  FC,
} from 'react';
import { AppContextProps, Step, Questions } from '../types';

const defaultState = {
  steps: { current: [] as Step[] },
  currentStep: 0,
  questions: [],
  answeredSteps: [] as boolean[],
  answers: [] as string[],
  nextStep: () => {
    throw new Error('nextStep function not implemented');
  },
  prevStep: () => {
    throw new Error('prevStep function not implemented');
  },
  answerQuestion: () => {
    throw new Error('answerQuestion function not implemented');
  },
  showResult: false,
};

export const AppContext = createContext<AppContextProps>(defaultState);

interface AppProviderProps {
  children: ReactNode;
}

export const AppProvider: FC<AppProviderProps> = ({ children }) => {
  const steps = useRef<Step[]>([
    { number: '1', hasRange: true, isInactive: false, asComplete: false },
    { number: '2', hasRange: true, isInactive: true, asComplete: false },
    { number: '3', hasRange: true, isInactive: true, asComplete: false },
    { number: '4', hasRange: true, isInactive: true, asComplete: false },
    { number: '5', hasRange: false, isInactive: true, asComplete: false },
  ]);

  const questions: Questions[] = [
    {
      question: 'Como você está se sentindo agora?',
      answers: [
        { text: 'Feliz' },
        { text: 'Triste' },
        { text: 'Ansioso' },
        { text: 'Bem animado' },
        { text: 'Entediado' },
      ],
      description: 'teste 123',
    },
    {
      question: 'Qual o seu estilo de filme favorito?',
      answers: [
        { text: 'Ação' },
        { text: 'Aventura' },
        { text: 'Animação' },
        { text: 'Comédia' },
        { text: 'Drama' },
        { text: 'Fantasia' },
        { text: 'SciFi' },
        { text: 'Thriller' },
        { text: 'Terror/Suspense' },
      ],
      description: 'teste 123',
    },
    {
      question: 'Seu segundo estilo de filme favorito?',
      answers: [
        { text: 'Ação' },
        { text: 'Aventura' },
        { text: 'Animação' },
        { text: 'Comédia' },
        { text: 'Drama' },
        { text: 'Fantasia' },
        { text: 'SciFi' },
        { text: 'Thriller' },
        { text: 'Terror/Suspense' },
      ],
      description: 'teste 123',
    },
    {
      question: 'Qual periodo de lançamento mais lhe convém?',
      answers: [
        { text: '2000' },
        { text: '1990' },
        { text: '1980' },
        { text: '1970' },
        { text: '1960' },
        { text: '1950' },
        { text: '1940' },
      ],
      description: 'teste 123',
    },
  ];

  const [currentStep, setCurrentStep] = useState<number>(0);

  const [showResult, setShowResult] = useState<boolean>(false);

  const [answeredSteps, setAnsweredSteps] = useState<boolean[]>([
    false,
    false,
    false,
    false,
    false,
  ]);
  const [answers, setAnswers] = useState<string[]>([]);

  const nextStep = useCallback(() => {
    setCurrentStep((prev) => Math.min(prev + 1, steps.current.length - 1));
  }, []);

  const prevStep = useCallback(() => {
    setCurrentStep((prev) => Math.max(prev - 1, 0));
  }, []);

  const answerQuestion = useCallback(
    (text: string) => {
      if (currentStep < questions.length - 1) {
        setAnswers((prev) => {
          const newAnswers = [...prev];
          newAnswers[currentStep] = text;
          return newAnswers;
        });
        setAnsweredSteps((prev) => {
          const newAnsweredSteps = [...prev];
          newAnsweredSteps[currentStep] = true;
          return newAnsweredSteps;
        });
        nextStep();
      } else {
        setAnswers((prev) => {
          const newAnswers = [...prev];
          newAnswers[currentStep] = text;
          return newAnswers;
        });
        setAnsweredSteps((prev) => {
          const newAnsweredSteps = [...prev];
          newAnsweredSteps[currentStep] = true;
          return newAnsweredSteps;
        });
        setShowResult(true);
      }
    },
    [currentStep, nextStep, questions.length],
  );

  return (
    <AppContext.Provider
      value={{
        steps,
        currentStep,
        questions,
        answeredSteps,
        answers,
        nextStep,
        prevStep,
        answerQuestion,
        showResult,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
