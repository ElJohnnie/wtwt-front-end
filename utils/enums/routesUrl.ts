export enum RoutesUrls {
  HOME = '/',
  RESULT = '/result',
  ERROR = '/error',
}

export type RouteUrl = RoutesUrls.RESULT | RoutesUrls.HOME | RoutesUrls.ERROR;
