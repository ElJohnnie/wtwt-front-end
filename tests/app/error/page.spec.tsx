import React, { act } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ErrorPage from '../../../app/error/page';
import { AppContext } from '../../../contexts/AppContext';
import { RoutesUrls } from '../../../utils/enums/routesUrl';

jest.mock('lottie-react', () => ({
  __esModule: true,
  default: () => <div>Lottie Mock</div>,
}));

const redirectMock = jest.fn();
const replaceMock = jest.fn();

jest.mock('../../../hooks/useNavigation/index.tsx', () => ({
  ...jest.requireActual('../../../hooks/useNavigation/index.tsx'),
  useNavigation: () => ({
    redirect: redirectMock,
    replace: replaceMock,
  }),
}));

const mockContextValue = {
  steps: {
    current: [
      { number: '1', hasRange: true, isInactive: false, asComplete: false },
      { number: '2', hasRange: true, isInactive: true, asComplete: false },
      { number: '3', hasRange: true, isInactive: true, asComplete: false },
      { number: '4', hasRange: true, isInactive: true, asComplete: false },
      { number: '5', hasRange: false, isInactive: true, asComplete: false },
    ],
  },
  currentStep: 0,
  questions: [
    {
      question: 'Como você está se sentindo agora?',
      answers: [
        { text: 'Feliz', value: 'happy' },
        { text: 'Triste', value: 'sad' },
        { text: 'Ansioso', value: 'anxious' },
        { text: 'Bem animado', value: 'excited' },
        { text: 'Entediado', value: 'bored' },
      ],
      description: 'teste 123',
    },
  ],
  answeredSteps: [true, true, true, true],
  answers: ['any', 'any', 'any', 'any'],
  nextStep: jest.fn(),
  prevStep: jest.fn(),
  answerQuestion: jest.fn(),
  showResult: false,
  resetState: jest.fn(),
};

describe('ErrorPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Renderizando a página de erro', () => {
    const { container } = render(
      <AppContext.Provider value={mockContextValue}>
        <ErrorPage />
      </AppContext.Provider>,
    );

    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('Clicando no botão de voltar para a página inicial', () => {
    const { getByTestId } = render(
      <AppContext.Provider value={mockContextValue}>
        <ErrorPage />
      </AppContext.Provider>,
    );

    screen.logTestingPlaygroundURL();

    const backButton = getByTestId('go-to-start');
    expect(backButton).toBeInTheDocument();

    act(() => {
      fireEvent.click(backButton);
    });

    expect(replaceMock).toHaveBeenCalledWith(RoutesUrls.HOME);
  });

  it('Clicando no botão de tentar novamente', () => {
    const { getByTestId } = render(
      <AppContext.Provider value={mockContextValue}>
        <ErrorPage />
      </AppContext.Provider>,
    );

    const tryAgainButton = getByTestId('try-again');
    expect(tryAgainButton).toBeInTheDocument();

    act(() => {
      fireEvent.click(tryAgainButton);
    });

    expect(replaceMock).toHaveBeenCalledWith(RoutesUrls.RESULT);
  });
});
