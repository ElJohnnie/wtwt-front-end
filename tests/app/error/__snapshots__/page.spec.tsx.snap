// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage Renderizando a p√°gina de erro 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center min-h-screen p-4"
  >
    <div
      class="p-8 md:p-16 max-w-lg w-full text-center"
    >
      <h1
        class="text-4xl font-bold text-pink-600 mb-4"
      >
        Algo saiu do controle
      </h1>
      <p
        class="text-pink-700 mb-8"
      >
        Gostaria de tentar novamente?
      </p>
      <div
        class="my-4 flex space-x-4 p-4"
      >
        <div
          class="w-full flex flex-col justify-center items-center"
        >
          <button
            class="w-24 h-24 rounded-full   inline-flex items-center justify-center   text-white text-[24px] font-bold bg-pink-600 hover:bg-pink-800"
          >
            <svg
              class="size-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16.023 9.348h4.992v-.001M2.985 19.644v-4.992m0 0h4.992m-4.993 0 3.181 3.183a8.25 8.25 0 0 0 13.803-3.7M4.031 9.865a8.25 8.25 0 0 1 13.803-3.7l3.181 3.182m0-4.991v4.99"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <p
            class="text-[18px] mt-1 text-pink-600"
          >
            Tentar novamente
          </p>
        </div>
        <div
          class="w-full flex flex-col justify-center items-center"
        >
          <button
            class="w-24 h-24 rounded-full   inline-flex items-center justify-center   text-white text-[24px] font-bold bg-pink-600 hover:bg-pink-800"
          >
            <svg
              class="size-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.25 9V5.25A2.25 2.25 0 0 1 10.5 3h6a2.25 2.25 0 0 1 2.25 2.25v13.5A2.25 2.25 0 0 1 16.5 21h-6a2.25 2.25 0 0 1-2.25-2.25V15m-3 0-3-3m0 0 3-3m-3 3H15"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <p
            class="text-[18px] mt-1 text-pink-600"
          >
            Refazer novamente
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
